lib/msal-core/src/utils/WindowUtils.ts

```diff
@@ -0,0 +1,202 @@
+
+                resolve(frameHandle);
+            }, timeoutMs);
+        });
+    }
+
+    /**
+     * @hidden
+     * Adds the hidden iframe for silent token renewal.
+     * @ignore
+     */
+    static addHiddenIFrame(iframeId: string, logger: Logger): HTMLIFrameElement {
+        if (typeof iframeId === "undefined") {
+            return null;
+        }
+
+        logger.info("Add msal frame to document:" + iframeId);
+        let adalFrame = document.getElementById(iframeId) as HTMLIFrameElement;
+        if (!adalFrame) {
+            if (document.createElement &&
+        document.documentElement &&
+        (window.navigator.userAgent.indexOf("MSIE 5.0") === -1)) {
+                const ifr = document.createElement("iframe");
+                ifr.setAttribute("id", iframeId);
+                ifr.style.visibility = "hidden";
+                ifr.style.position = "absolute";
+                ifr.style.width = ifr.style.height = "0";
+                ifr.style.border = "0";
+                ifr.setAttribute("sandbox", "allow-same-origin");
+                adalFrame = (document.getElementsByTagName("body")[0].appendChild(ifr) as HTMLIFrameElement);
+            } else if (document.body && document.body.insertAdjacentHTML) {
+                document.body.insertAdjacentHTML("beforeend", "<iframe name='" + iframeId + "' id='" + iframeId + "' style='display:none'></iframe>");
+            }
+
+            if (window.frames && window.frames[iframeId]) {
+                adalFrame = window.frames[iframeId];
+            }
+        }
+
+        return adalFrame;
+    }
+
+    /**
+     * @hidden
+     * Find and return the iframe element with the given hash
+     * @ignore
+     */
+    static getIframeWithHash(hash: string) {
+        return Array.from(document.getElementsByTagName("iframe")).find(iframe => {
+            try {
+                return iframe.contentWindow.location.hash === hash;
+            } catch (e) {
+                return false;
+            }
+        });
+    }
+
+    /**
+     * @hidden
+     * Returns an array of all the popups opened by MSAL
+     * @ignore
+     */
+    static getPopups(): Array<Window> {
+        if (!window.openedWindows) {
+            window.openedWindows = [];
+        }
+
+        return window.openedWindows;
+    }
+
+    /**
+     * @hidden
+     * Find and return the popup with the given hash
+     * @ignore
+     */
+    static getPopUpWithHash(hash: string) {
+        return WindowUtils.getPopups().find(popup => {
+            try {
+                return popup.location.hash === hash;
+            } catch (e) {
+                return false;
+            }
+        });
+    }
+
+    /**
+     * @hidden
+     * Add the popup to the known list of popups
+     * @ignore
+     */
+    static trackPopup(popup: Window) {
+        WindowUtils.getPopups().push(popup);
+    }
+
+    /**
+     * @hidden
+     * Close all popups
+     * @ignore
+     */
+    static closePopups() {
```

Comment:
would there ever be anything else in the openedWindows than popups?

Reply:
We're gonna hopefully remove `openedWindows` soon.